src/function/matrix/eigs/complex.js

```diff
@@ -0,0 +1,316 @@
+export function createComplex({addScalar, subtract, multiply, sqrt, abs, bignumber, diag, qr}) {
+    /**
+     * @param {number[][]} arr
+     * @param {number} N
+     * @param {number|BigNumber} prec
+     * @param {'number'|'BigNumber'|'Complex'} type
+     */
+    function main(arr, N, prec, type)
+    {
+        // TODO implement QR for complex matrices
+        if (type === 'Complex')
+            throw new TypeError('Complex matrices not yet supported')
+
+
+        // TODO check if any row/col are zero except the diagonal
+
+        let R = balance(arr, N, prec, type)
+
+        // TODO if magnitudes of elements vary over many orders,
+        // move greatest elements to the top left corner
+
+        reduceToHessenberg(arr, N, prec, type, R)
+
+        let values = iterateUntilDiagonal(arr, N, prec, type, R)
+
+        // TODO find vectors
+
+        return { values }
+    }
+
+    /**
+     * @param {number[][]} arr
+     * @param {number} N
+     * @param {number} prec
+     * @param {'number'|'BigNumber'|'Complex'} type
+     * @returns {number[][]}
+     */
+    function balance(arr, N, prec, type)
+    {
+        const big  = type === 'BigNumber'
+        const cplx = type === 'Complex'
+
+        const bigZero = bignumber(0)
+        const bigOne  = bignumber(1)
+
+        // base of the floating-point arithmetic
+        const radix = big ? bignumber(10) : 2
+        const radixSq = multiply(radix, radix)
+
+        // the diagonal transformation matrix R
+        const Rdiag = Array(+N).fill(big ? bigOne : 1)
+
+        // this isn't the only time we loop thru the matrix...
+        let last = false
+
+        while (!last)
```

Comment:
Is there a risk of getting into an infinite loop in this `while` loop?

Reply:
I'm quite certain that no. The loop iterates through every row-column pair only once, then they are balanced for good. The only reason for using `while` instead of `for (let i = 0; i < N; i++)` is that the rows and columns get switched, so there's no clear ordering.